File: ride/src/main/scala/com/shevchyk/ride/domain/RideRating.scala
package com.shevchyk.ride.domain

import com.shevchyk.core.domain.*
import zio.json.*
import java.time.Instant
import java.util.UUID
import com.github.f4b6a3.uuid.UuidCreator

case class RideRatingId(value: UUID) derives JsonCodec

object RideRatingId:
  def generate(): RideRatingId = RideRatingId(UuidCreator.getTimeOrderedEpoch())

final case class RideRating(
    id: RideRatingId,
    rideId: RideId,
    clientId: PersonId,
    driverId: PersonId,
    companyId: CompanyId,
    rating: Int,
    comment: Option[String] = None,
    createdAt: Instant = Instant.now()
) derives JsonCodec

final case class CreateRatingRequest(
    // Optional: the ride id is taken from the URL path, so clients only send
    // `rating` and `comment`. Keeping it optional avoids a JSON-decode failure.
    rideId: Option[String] = None,
    rating: Int,
    comment: Option[String] = None
) derives JsonCodec

File: web/lib/modules/core/models/ride_rating.dart
import '../json_parse.dart';

class RideRating {
  final String id;
  final String rideId;
  final String clientId;
  final String driverId;
  final String companyId;
  final int rating;
  final String? comment;
  final DateTime createdAt;

  const RideRating({
    required this.id,
    required this.rideId,
    required this.clientId,
    required this.driverId,
    required this.companyId,
    required this.rating,
    this.comment,
    required this.createdAt,
  });

  factory RideRating.fromJson(Map<String, dynamic> json) {
    return RideRating(
      id: json['id'] ?? '',
      rideId: json['rideId'] ?? '',
      clientId: json['clientId'] ?? '',
      driverId: json['driverId'] ?? '',
      companyId: json['companyId'] ?? '',
      rating: json['rating'] ?? 0,
      comment: json['comment'],
      createdAt: JsonParse.requiredDateTime(json, 'createdAt'),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'rideId': rideId,
      'clientId': clientId,
      'driverId': driverId,
      'companyId': companyId,
      'rating': rating,
      'comment': comment,
      'createdAt': createdAt.toIso8601String(),
    };
  }
}
